src/env-inl.h

```diff
@@ -344,8 +344,35 @@ inline void Environment::RegisterHandleCleanup(uv_handle_t* handle,
   handle_cleanup_queue_.push_back(HandleCleanup{handle, cb, arg});
 }
 
-inline void Environment::FinishHandleCleanup(uv_handle_t* handle) {
-  handle_cleanup_waiting_--;
+template <typename T, typename OnCloseCallback>
+inline void Environment::CloseHandle(T* handle, OnCloseCallback callback) {
+  handle_cleanup_waiting_++;
+  static_assert(sizeof(T) >= sizeof(uv_handle_t), "T is a libuv handle");
+  static_assert(offsetof(T, data) == offsetof(uv_handle_t, data),
+                "T is a libuv handle");
+  static_assert(offsetof(T, close_cb) == offsetof(uv_handle_t, close_cb),
+                "T is a libuv handle");
+  struct CloseData {
+    Environment* env;
+    OnCloseCallback callback;
+    void* original_data;
+  };
+  handle->data = new CloseData { this, callback, handle->data };
+  uv_close(reinterpret_cast<uv_handle_t*>(handle), [](uv_handle_t* handle) {
```

Comment:
Question: why not use a `[&]` lambda instead of the explicit closure struct?
Anyway, a line or two of comments would be nice:
```C++
// This struct is used to close over these three values, and is passed back tough the handle->data pointer.
// It is stored on the heap and is freed by std::unique_ptr
```
---
I would suggest using a different name for the `T* handle` and the `uv_handle_t* handle`.

Reply:
> Question: why not use a `[&]` lambda instead of the explicit closure struct?

Because that doesn’t work, it’s not convertible to the C signature that libuv needs

Reply:
Ack.

Reply:
So will you be willing to add a comment, since it's a bit of non trivial code + your last comment that a `[&]` lambda doesn't work

Reply:
@refack Yeah, sorry, should have been clearer about that – I’ll add a comment. :)

Fwiw, we’re also running into this issue in a lot of places in core – capturing C++ lambdas and the libuv API just don’t mix very well. I think it’s fine to leave some short comments, but ultimately, it’s just a language-level thing…